### ISSUE TITLE
Is there a way to set Rake.application.options.dryrun to true when running commands from terminal with --dry-run option?

### ISSUE THREAD

**Issue by bogdanRada** [capistrano/capistrano#1462]
I have a bit of a problem with dry-run option. My tasks are getting executed even if i run the commands with --dry-run options

and trying to debug i saw that Rake.application.options.dryrun is set to nil  and that makes Rake to not respect the --dry-run option .

Is there a way to fix this? would be really great :)

**Comment by leehambley** [comment]
What versions of everything are you using, and what command are you running?

**Comment by bogdanRada** [comment]
it works with capistrano 3.2.1, but if i use a higher version, even latest version it doesn't work, I belive this happens because of this commit: https://github.com/capistrano/capistrano/commit/d868b95d034aa2c5c184759bd52ec1ed152eaeac 

i was working on this project, https://github.com/bogdanRada/capistrano-gitflow ( this is compatible with capistrano 3 and capistrano 2..the readme is not updated however) and i tried to run my tasks with --dry-run options and i saw that i am being asked about the tag to use for release, which shouldn't happen and this is how i realized this problem.

**Comment by bogdanRada** [comment]
i am running  "bundle exec cap gitflow:calculate_tag --dry-run --trace " and if i print Rake.application.options i see there is only "trace" and "backtrace" set to true, but "dryrun" is not set to true.

**Comment by leehambley** [comment]
Can you try with that patch reverted, if you can get it fixed, I'll merge a
patch, otherwise it'll probably not be fixable in the next days.

Lee Hambley
http://lee.hambley.name/
+49 (0) 170 298 5667

**Comment by bogdanRada** [comment]
Yes, i checked, that seems to work but if i revert it will override the capistrano settings for dry-run option

But i think the current version can be used like this, if i add few lines in Capistrano::Application class like this: 

```ruby
    def sort_options(options)
      not_applicable_to_capistrano = %w(quiet silent verbose)
      options.reject! do |(switch, *)|
        switch =~ /--#{Regexp.union(not_applicable_to_capistrano)}/
      end
      
       @original_options = options.clone 
      super.push(version, dry_run, roles, hostfilter)
    end

   def dry_run
      ['--dry-run', '-n',
        "Do a dry run without executing actions",
        lambda { |value|
          Configuration.env.set(:sshkit_backend, SSHKit::Backend::Printer)
          original_option = @original_options.detect{|elem| elem[0].include?('--dry-run')}
          original_option[3].call value if original_option
        }
      ]
    end
``

This will also set the option dryrun to true. I checked and it works. Not sure if it best solution. Let me know what you think.  Maybe there is a better way of doing this.

**Comment by leehambley** [comment]
Thanks for the patch, I'll have a look at integrating it when I have time.

> However what is annoying is that having this option set to true, it does not follow the invocation chain of the task , but that is something that can be fixed in rake gem i think

What do you mean, exactly?

**Comment by bogdanRada** [comment]
First of all, thanks for such a prompt response. I really appreciate the help. 

The other thing i said, it's not actually related to capistrano, but more to rake gem

i have another project: https://github.com/bogdanRada/capistrano_multiconfig_parallel and i was trying to show progress of tasks while they are running in parallel, but in order to do that i run them with option "--dry-run" first to save the invocation chain  for each worker and then while they are executing i can show progress.

However now I  discovered this bug with '--dry-run' that actually runs the tasks, and i can't really do this using current version (because that will execute some of the tasks on server)  

So if i revert completely the patch or use my patch and try to run this command:

```
 bundle exec cap ghcp-api:development deploy --dry-run --trace
```
 I get only this output and then it stops:

```
** Invoke ghcp-api:development (first_time)
** Execute (dry run) ghcp-api:development
** Invoke deploy (first_time)
** Invoke gitflow:verify_up_to_date (first_time)
** Execute (dry run) gitflow:verify_up_to_date
** Execute (dry run) deploy
```

But probably this thing with rake and hooks is most likely a bug in rake gem or something . So you can
deconsider this afirmation:
  
> However what is annoying is that having this option set to true, it does not follow the invocation chain
> of the task , but that is something that can be fixed in rake gem i think

**Comment by leehambley** [comment]
Part of the problem is that, in an ideal world, Rake would be driven by it's prerequisites, solely. We would be able to expect an up-to-date remote directory, and work backwards from there, create a release, create the cache, etc, etc, etc all the way back to asking for input, etc.

Regrettably most people just don't think, and it's incredibly difficut to get *remote* prerequisites to rebuild sanely (i.e differences in clocks).

To make a long story short, some tasks have to run, in order to [`Rake::Task#invoke`](http://ruby-doc.org/stdlib-1.9.3/libdoc/rake/rdoc/Rake/Task.html#method-i-invoke) to start their next tasks, if the task is running under `--dry-run`, the task never runs, so it's "after" tasks (which logically should be tasks which depend on this task as prerequisites) are never called.

If that's any way unclear, I would be happy to write a more thorough explanation, but I think for waht you plan, to gauge "progress' by counting the number of tasks by dry-running, and making an overview, you will struggle, the best thing I can suggest, is maybe to… implement a monkey patch, and somewhere hidden on disk, cache the number and names of tasks, and on the first-run/cold-cache you have to simply show an unknown progress (or something)… and, maybe blow away the cache if your list of tasks this time, differs with what was in the cache.

Finally `pv` has a nice concept for this - http://www.ivarch.com/programs/pv.shtml note in the example screenshot that pv doesn't know how large the 2nd and 3rd files are, so it can't show a % progress, but it can show elapsed time, and some other data.

**Comment by bogdanRada** [comment]
wow. that is awesome. thanks. i am definitely taking this in consideration. i wil most likely do what you suggested. i think is a good idea. thank you very much

**Comment by leehambley** [comment]
Happy to help, I'll leave this issue that our dry-run is broken open, that's worth fixing, even if it doesn't directly help you. `pv` is a wonderful little tool, if one ever remembers to use it :) Keep me informed how your multi-capistrano thing is running, it sounds really interesting.

**Comment by bogdanRada** [comment]
ok. thanks :)

**Comment by phallstrom** [comment]
:+1:

**Comment by rainchen** [comment]
+1 for fixing the `Rake.application.options.dry_run`, I need to add some custom recipes, need a place  to check if cap is running with --dry-run options. Using `self.is_a?(SSHKit::Backend::Printer)` is working but weird.

**Comment by mattbrictson** [comment]
Just to make sure I understand the problem here:

`--dry-run` means different things for Rake and for Capistrano:

* **Rake** interprets a dry run as just printing tasks, but not executing the bodies of those tasks. This works for prerequisite-driven Rakefiles where all tasks are known at load time.
* **Capistrano** does *not* do a Rake dry run. Instead, it *actually executes all tasks*, but uses a `Printer` for the SSHKit backend rather than the default SSH. The result is that any `execute` will just print instead of running the remote command. So Capistrano's dry run is still a dry run of sorts, but different than what Rake does.

Am I on the right track here?

Then, the questions are:

1. Is `cap --dry-run` "working as designed"? (Then this issue can be closed.)
2. Is there a need to support *both* styles of dry-run? (Then I suggest a new feature request issue be created.)

> need a place to check if cap is running with --dry-run options

@rainchen I think your feature request is a good one, but could you open it as a separate GitHub issue? Maybe we could expose a `dry_run?` method in the Capistrano DSL to address that.

**Comment by phallstrom** [comment]
@mattbrictson That sounds like the right track to me. My case is that I maintain slackistrano. A gem to post deploy notifications to Slack. My problem is if users do a dry run those notifications are still posted. So I need a way to know if we're in a dry run so I can do what SSH does and simply print "would have posted to a slack channel", etc.

I don't think we need to support both styles. At least I don't. Being able to check if we're in a dry run and alter course is good enough for me.

**Comment by mattbrictson** [comment]
> Being able to check if we're in a dry run and alter course is good enough for me

Cool. I'll open an issue then… done: #1564

**Comment by bogdanRada** [comment]
Yes i agree Rake and Capistrano do different things. But this used to work in versions less or equal to 3.2.1 . In higher versions though it only runs 3 tasks or so and then it just exits abnormally. Hence this issue. I already posted above that it exits abnormally and also posted a solution for it which i tested and it does fix the problem. might not be the best solution but i am looking though for a solution. I am not asking that capistrano should do what rake does. I only want to make capistrano work just like it used to work in versions less or equal to 3.2.1. Could someone help?

**Comment by mattbrictson** [comment]
@bogdanRada Sorry, I am confused: this is a long thread and I guess I haven't gotten to the core of your bug report. Could you change the title of this issue to be a concise description of the bug? Right now it reads like a feature request.

Please understand that I am trying to unravel this thread which seems to be conflating multiple bug reports and feature requests. I'd like to focus this discussion on the 3.4.0 "exits abnormally" bug and split off feature requests into separate issues.

You said:

> if i ... use my patch and try to run [`cap --dry-run`]  ... I get only this output and then it stops

I interpreted this as meaning that *your patch* is the one that is causing Capistrano to exit prematurely. Did I read that incorrectly?

If Capistrano 3.4.0 without any patching exhibits the "only runs about 3 tasks and exits abnormally" behavior, then that is a bug I would definitely like to investigate. Can you provide some more details on that?

**Comment by mattbrictson** [comment]
Also, one more clarification: the `--dry-run` behavior of 3.4.0 (as I described [previously](https://github.com/capistrano/capistrano/issues/1462#issuecomment-173059474)) is how Capistrano was designed to work. The code for this behavior has been in place since 3.0.0 as you can see in commit 35af438.

The 3.2.1 and prior behavior you are citing was unintentional. Due to a bug, the Capistrano dry run option was being ignored and Rake's dry run option was being used instead. This bug was corrected in commit d868b95.

Here's how I propose we move forward:

1. I will close this issue because it has become a mess of many entangled requests.
2. I will open a bug about the dry-run deployment process exiting prematurely.
3. I will open a feature request that Capistrano have a Rake-style dry run mode (like 3.2.1 did) where the tasks are printed but not executed.

Is this OK with you?

**Comment by bogdanRada** [comment]
Hello thanks for your prompt response. 
This happens with any version higher than 3.2.1 and i did made a mistake when i said "if i run my patch". i admit english is not my first language and at that time i was in a hurry so i think i wrote that by mistake. 

what i probay wanted to say is if i revert completely this patch (commit)  #d868b95d034aa2c5c184759bd52ec1ed152eaeac.  or run master branch i get same result. 
  using my patch does seem to work although i admit is not best solution and does seem a bit hacky and i was looking for a cleaner solution. 

 my gemfile had only this gems: capistrano, capistrano-gitflow , rvm-capistrano, capistrano-bundler . and i was using Rake 10.4
  not sure if this issue is from capistrano or from rake though.

**Comment by bogdanRada** [comment]
sorry @mattbrictson i only saw your last comment now. 

yes i ok with that. thank you very much

**Comment by mattbrictson** [comment]
OK, great. Thanks for your patience.

Closing in favor of #1565 and #1566.
